### PR TITLE
cssSafeName: Replace nonCssSafeCharacters with encoded character

### LINKED ISSUE #20: Duplicate class names for different Pangolineages
Flagged by @corneliusroemer on [Slack](https://bedfordlab.slack.com/archives/C036L7V615F/p1713437813749909). 

Pango lineages that are identical after stripping the periods from the variant names result in showing both lines on hover of legend. 

Example:  JN.1.43 and JN.1.4.3

image
image (1)

This is due to running the variant names through `cssSafeName`

https://github.com/nextstrain/forecasts-viz/blob/33f698128c8387e0eb71638a467e3c97a4382c95/src/lib/utils/d3Graph.js#L117

which strip nonCssSafeCharacters

https://github.com/nextstrain/forecasts-viz/blob/33f698128c8387e0eb71638a467e3c97a4382c95/src/lib/utils/cssSafeName.js#L17-L20

### PR DESCRIPTION
**Pull request by joverlee521**
~Fixes the bug where stripped class names were clashing where Pango lineage variant names were identical after stripping the `.`~

~However, this is not entirely safe since we can still get name clashes in the future. E.g. `class!name` and `class.name` both get replaced with `class_name`.~

Replaces unsafe characters with the hexadecimal ASCII value that is
surrounded by underscores so that we don't run into name clashes.

Fixes the bug where stripped class names were clashing where Pango
lineage variant names were identical after stripping the `.`

Script taken from https://github.com/nextstrain/forecasts-viz/pull/21#issuecomment-2064981508

## Related issue(s)

Resolves #20 

## Checklist

- [ ] Checks pass

### CODE REVIEW

**Comment by joverlee521** [comment]
> However, this is not entirely safe since we can still get name clashes in the future.

Not sure how to get around this since we need reproducible names that can be accessed later in the app

https://github.com/nextstrain/forecasts-viz/blob/33f698128c8387e0eb71638a467e3c97a4382c95/src/lib/utils/d3Graph.js#L258

**Comment by corneliusroemer** [comment]
Is there no way to properly escape such characters? Could also hash it with md5 and use the hash as classname?

This is what[ ChatGPT4 suggested](https://chat.openai.com/share/196ba695-0ad8-4a58-a5e6-6fd1c482093a), using hex character encoding:

```js
const nonCssSafeCharacters = /[!"#$%&'()*+,./:;<=>?@[\\\]^`{|}~\s]/g;

const invalidBeginningOfClassname = /^([0-9]|--|-[0-9])/;

function encodeCharacter(c) {
  return '_' + c.charCodeAt(0).toString(16) + '_';
}

export const cssSafeName = str => {
  if (typeof str !== 'string') {
    return '';
  }

  let encodedClassName = str.replace(nonCssSafeCharacters, encodeCharacter);

  if (invalidBeginningOfClassname.test(encodedClassName)) {
    encodedClassName = `_${encodedClassName}`;
  }

  return encodedClassName;
};
```

File: src/lib/utils/cssSafeName.js [hunk 1]
```diff
@@ -1,25 +1,57 @@
...
   }
 
-  const strippedClassname = str.replace(
-    new RegExp(nonCssSafeCharacters, 'g'),
-    ''
-  )
+  let encodedClassName = str.replace(nonCssSafeCharacters, encodeCharacter);
+
+  if (invalidBeginningOfClassname.test(encodedClassName)) {
+    encodedClassName = `_${encodedClassName}`;
```
Comment (corneliusroemer) [inline src/lib/utils/cssSafeName.js]: One could return here directly

```suggestion
    return `_${encodedClassName}`;
```

Comment (corneliusroemer) [reply src/lib/utils/cssSafeName.js]: actually the original was best I think :)

Comment (joverlee521) [reply src/lib/utils/cssSafeName.js]: Updated in https://github.com/nextstrain/forecasts-viz/pull/21/commits/e5f3d127f83159d05b7c511578eaef9c53f006e1

**Comment by joverlee521** [comment]
I deployed the PR to GH Pages with `npm run deploy` on this branch. 

JN.1.43 and JN.1.4.3 are highlighted separately in https://nextstrain.github.io/forecasts-viz/?dataset=https://data.nextstrain.org/files/workflows/forecasts-ncov/gisaid/pango_lineages/global/mlr/latest_results.json

**Comment by joverlee521** [comment]
Merging and vendoring to forecasts-ncov.

**Pull request merged**